File: ConcreteApp/app/src/main/java/com/example/brunovsiq/concreteapp/models/RepositoryDataSourceFactory.java
package com.example.brunovsiq.concreteapp.models;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;
import io.reactivex.subjects.PublishSubject;

public class RepositoryDataSourceFactory extends DataSource.Factory {

    private MutableLiveData<RepositoryDataSource> mutableLiveData;
    private RepositoryDataSource dataSource;

    public RepositoryDataSourceFactory() {
        mutableLiveData = new MutableLiveData<>();
    }

    @Override
    public DataSource create() {

        dataSource = new RepositoryDataSource();
        mutableLiveData.postValue(dataSource);
        return dataSource;
    }
}

File: app/src/main/java/com/joaoibarra/gitapp/api/paging/RepositoryDatasourceFactory.kt
package com.joaoibarra.gitapp.api.paging

import android.arch.paging.DataSource
import com.joaoibarra.gitapp.api.GitApi
import com.joaoibarra.gitapp.api.model.Repo
import io.reactivex.disposables.CompositeDisposable

class RepositoryDatasourceFactory (
    private val compositeDisposable: CompositeDisposable,
    private val gitApi: GitApi
): DataSource.Factory<Int, Repo>() {

    override fun create(): DataSource<Int, Repo> {
        return RepositoryDataSource(gitApi, compositeDisposable)
    }
}
